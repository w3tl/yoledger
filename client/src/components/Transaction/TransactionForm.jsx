import React from 'react';
import PropTypes from 'prop-types';
import { Button, DateInput } from '../Elements';
import { AccountInput } from '../Account';
import AmountInput from '../AmountInput';
import TransactionTypeSelect from './TransactionTypeSelect';
import { transactionTypePropType, transactionPropType } from './propTypes';

class TransactionForm extends React.Component {
  static propTypes = {
    transaction: transactionPropType,
    type: transactionTypePropType,
    onCreate: PropTypes.func,
    onSave: PropTypes.func,
    onDelete: PropTypes.func,
  }

  constructor(props) {
    super(props);
    const { transaction, type } = props;
    this.state = {
      isCreate: !transaction.id,
      amount: transaction.amount,
      source: transaction.source.name,
      destination: transaction.destination.name,
      date: transaction.date,
      type,
    };
  }

  handleChange = name => ({ target }) => {
    this.setState({
      [name]: target.value,
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const {
      isCreate, amount, source, destination, date,
    } = this.state;
    const { transaction, onSave, onCreate } = this.props;
    if (isCreate) {
      onCreate({
        amount, source, destination, date,
      });
    } else {
      onSave(transaction.id, {
        amount, source, destination, date,
      });
    }
  }

  handleDelete = () => {
    const { onDelete } = this.props;
    onDelete();
  }

  handleTypeChange = ({ target }) => {
    this.setState({ type: target.value });
  }

  render() {
    const {
      amount, source, destination, date, type, isCreate,
    } = this.state;

    return (
      <form onSubmit={this.handleSubmit}>
        <TransactionTypeSelect value={type} onChange={this.handleTypeChange} />
        <AccountInput required label="From" value={source} onChange={this.handleChange('source')} />
        <AccountInput required label="To" value={destination} onChange={this.handleChange('destination')} />
        <AmountInput required label="Amount" value={amount} onChange={this.handleChange('amount')} />
        <DateInput label="When" value={date} onChange={this.handleChange('date')} />
        {!isCreate && <Button id="delete" onClick={this.handleDelete}>Delete</Button>}
        {source && destination && (
          <Button type="submit">
            Save
          </Button>)}
      </form>
    );
  }
}

TransactionForm.defaultProps = {
  transaction: {
    id: null,
    amount: 0,
    source: {
      name: '',
    },
    destination: {
      name: '',
    },
    date: new Date().toISOString(),
  },
  type: 'expense',
  onCreate: () => {},
  onSave: () => {},
  onDelete: () => {},
};

export default TransactionForm;
