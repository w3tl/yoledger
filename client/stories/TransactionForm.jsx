import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { TransactionForm } from '../src/components';

const transaction = {
  amount: 10,
  source: { name: 'Cash' },
  destination: { name: 'Food' },
  date: new Date('2018-01-01').toString(),
};

storiesOf('Transaction/Form', module)
  .add('default', () => <TransactionForm onSave={action('onSave')} />)
  .add('with value', () => <TransactionForm {...transaction} />);
