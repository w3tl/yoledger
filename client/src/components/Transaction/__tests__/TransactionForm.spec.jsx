import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TransactionForm from '../TransactionForm';
import transactions from '../mockData';

configure({ adapter: new Adapter() });

describe('TransactionForm', () => {
  it('should render correctly', () => {
    const output = shallow(
      <TransactionForm transaction={transactions[0]} />,
    );
    expect(output.state('isCreate')).toBeFalsy();
    expect(output.find('Button')).toHaveLength(2);
    expect(output).toMatchSnapshot();
  });

  it('should render delete button when transaction prop have\'t id field', () => {
    const output = shallow(
      <TransactionForm />,
    );
    expect(output.state('isCreate')).toBeTruthy();
    expect(output.find('Button')).toHaveLength(1);
  });

  it('should handle submit button click', () => {
    const transaction = transactions[0];
    const onSave = jest.fn();
    const output = mount(
      <TransactionForm
        transaction={transaction}
        onSave={onSave}
      />,
    );
    output.find('form').simulate('submit');
    expect(onSave.mock.calls[0]).toMatchSnapshot();
  });
});
