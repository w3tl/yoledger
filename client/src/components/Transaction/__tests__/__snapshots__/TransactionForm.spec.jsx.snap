// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionForm should handle submit button click 1`] = `
Array [
  Object {
    "amount": 10,
    "date": "2018-01-01T00:00:00.000Z",
    "destination": "Food",
    "source": "Cash",
  },
]
`;

exports[`TransactionForm should render correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <Component
    onChange={[Function]}
    value="expense"
  />
  <Component
    label="From"
    onChange={[Function]}
    required={true}
    value="Cash"
  />
  <Component
    label="To"
    onChange={[Function]}
    required={true}
    value="Food"
  />
  <Component
    label="Amount"
    onChange={[Function]}
    required={true}
    value={10}
  />
  <Component
    label="When"
    onChange={[Function]}
    value="2018-01-01T00:00:00.000Z"
  />
  <Button
    id="delete"
    onClick={[Function]}
    type="button"
  >
    Delete
  </Button>
  <Button
    type="submit"
  >
    Save
  </Button>
</form>
`;
